src/core/symbology/qgscategorizedsymbolrenderer.cpp

```diff
@@ -668,6 +668,39 @@ QgsFeatureRenderer *QgsCategorizedSymbolRenderer::create( QDomElement &element,
   QgsSymbolMap symbolMap = QgsSymbolLayerUtils::loadSymbols( symbolsElem, context );
   QgsCategoryList cats;
 
+  // Value from string (long, ulong, double and string)
+  const auto valueFromString = []( const QString & value, const QString & valueType ) -> QVariant
+  {
+    if ( valueType == QStringLiteral( "double" ) )
```

Comment:
Can you explain the rationale behind this change?

Reply:
Yes, this is the core change: when reading a stored category value from the project XML there was no way to know if the value was a string or an int or a double (we also need to handle the case of complex expressions so there is no way to rely on the field type, which wasn't stored either).

I couldn't find any other solution than to store the value type in the XML and use it when reading back the value at project load time. This works well even if the category value comes from a complex expression.

The main reason why we need to know the category value type is to get the correct widget in the (new) delegate, this way we can use a QgsDoubleSpinBox with the correct min/max and decimal places.